lrauv_ignition_plugins/src/ScienceSensorsSystem.cc

```diff
@@ -606,6 +710,249 @@ void ScienceSensorsSystemPrivate::PclVectorToEigen(
+  return a.x < b.x && a.y < b.y && a.z < b.z;
+}
+
+/////////////////////////////////////////////////
+void ScienceSensorsSystemPrivate::FindTrilinearInterpolators(
+  pcl::PointXYZ &_pt,
+  std::vector<int> &_inds,
+  std::vector<float> &_sqrDists,
+  std::vector<pcl::PointXYZ> &_interpolators1,
+  std::vector<pcl::PointXYZ> &_interpolators2)
+{
+  // Sanity checks
+  // Vector not populated
+  if (this->timeSpaceCoords.size() == 0)
+  {
+    return;
+  }
+  // Point cloud not populated
+  if (this->timeSpaceCoords[this->timeIdx]->size() == 0)
+  {
+    return;
+  }
+  if (_inds.size() == 0 || _sqrDists.size() == 0)
+  {
+    ignwarn << "FindInterpolators(): Invalid neighbors aray size ("
+            << _inds.size() << " and " << _sqrDists.size()
+            << "). No neighbors to use for interpolation. Returning NaN."
+            << std::endl;
+    return;
+  }
+  if (_inds.size() != _sqrDists.size())
+  {
+    ignwarn << "FindInterpolators(): Number of neighbors != number of "
+            << "distances. Invalid input. Returning NaN." << std::endl;
+    return;
+  }
+
+  // Two slices of different depth z values
+  pcl::PointCloud<pcl::PointXYZ> zSlice1, zSlice2;
+  // Indices of points in the z slices
+  std::vector<int> zSliceInds1, zSliceInds2;
+
+  // Indices and distances of neighboring points in the search results
+  // 4 above, 4 below
+  std::vector<int> interpolatorInds1, interpolatorInds2;
+  std::vector<float> interpolatorSqrDists1, interpolatorSqrDists2;
+
+  // Step 1: 1st NN, in its z slice, search for 4 NNs
+
+  // 1st nearest neighbor
+  // The distances from kNN search are sorted, so can always just take [0]th
+  int nnIdx = _inds[0];
+  float minDist = _sqrDists[0];
+  // Get z of neighbor
+  float nnZ = this->timeSpaceCoords[this->timeIdx]->at(nnIdx).z;
+
+  // Debug output
+  igndbg << this->timeSpaceCoords[this->timeIdx]->size()
+    << " points in full cloud" << std::endl;
+
+  // Search in z slice for 4 nearest neighbors in this slice
+  this->CreateDepthSlice(nnZ, *(this->timeSpaceCoords[this->timeIdx]), zSlice1,
+    zSliceInds1);
+  igndbg << "1st nn idx " << nnIdx << ", dist " << sqrt(minDist)
+    << ", z slice " << zSlice1.points.size() << " points" << std::endl;
+  this->CreateAndSearchOctree(_pt, zSlice1,
+    interpolatorInds1, interpolatorSqrDists1, _interpolators1);
+  if (interpolatorInds1.size() < 4 || interpolatorSqrDists1.size() < 4)
+  {
+    ignwarn << "Could not find enough neighbors in 1st slice z = " << nnZ
+      << " for trilinear interpolation." << std::endl;
+    return;
+  }
+
+  // Step 2: exclude z slice of 1st NN from further searches.
+
+  // Remove all points in the z slice of the 1st NN, so that the 2nd NN will be
+  // found in another z slice.
+  // Set invert flag to get all but the depth slice.
+  pcl::PointCloud<pcl::PointXYZ> cloudExceptZSlice1;
+  std::vector<int> indsExceptZSlice1;
+  this->CreateDepthSlice(nnZ, *(this->timeSpaceCoords[this->timeIdx]),
```

Comment:
That saves performance, but it does add more code. f3f72c2

We need a `ExtractIndices` filter to get the inverted sub-cloud. It's less clean to read, because [ExtractIndices::setIndices()](http://docs.ros.org/en/hydro/api/pcl/html/classpcl_1_1ExtractIndices.html) only accepts PCL `IndicesPtr` type. It doesn't accept the `std::vector` returned from [PassThrough](http://docs.ros.org/en/hydro/api/pcl/html/classpcl_1_1PassThrough.html). So we have to convert the type first.

I tested the new replacement with the 3 csv files in `buoyant_tethys.sdf` and verified the output looks the same before and after this change. But again, the outputs from this PR are not correct, so I would not trust things beyond small mechanical changes.